File: docs/caching/method_dispatch_bench.rb
#!/usr/bin/env ruby
# Micro-benchmark: generic function dispatch (Ruby version)
# Basic dispatch performance test using a dispatch table

# Dispatch table
dispatch_table = {
  :long => lambda { |x| [:long, x] },
  :string => lambda { |x| [:string, x.length] },
  :array => lambda { |x| [:array, x.length] },
  :hash => lambda { |x| [:hash, x.length] },
  :symbol => lambda { |x| [:symbol, x] },
  :other => lambda { |x| [:other, x] }
}

def dispatch_clj_impl(x, dispatch_table)
  key = case x
        when Integer, Float then :long
        when String then :string
        when Array then :array
        when Hash then :hash
        when Symbol then :symbol
        else :other
        end
  dispatch_table[key].call(x)
end

# Test data: 200,000 calls with 5-type repeating cycle
test_data = []
cycle = [1, "test string", [1, 2, 3, 4, 5], { a: 1 }, :symbol]
200000.times { |i| test_data << cycle[i % 5] }

def benchmark_dispatch(iterations, test_data, dispatch_table)
  times = []
  iterations.times do |run|
    result = 0
    start_time = Time.now
    test_data.each do |item|
      r = dispatch_clj_impl(item, dispatch_table)
      result += 1 if r
    end
    elapsed = (Time.now - start_time) * 1000.0
    puts "  Run #{run + 1}: #{sprintf('%.1f', elapsed / 1000.0)} seconds"
    times << result
  end
  times
end

def run_all_benchmarks(test_data, dispatch_table)
  puts "\n================================"
  puts "Ruby Generic Function Dispatch Micro-Benchmark"
  puts "================================"
  puts "Implementation: Ruby #{RUBY_VERSION}"
  puts "Test data: 200,000 calls over repeating 5-type cycle"
  puts "  Type cycle: integer -> string -> array -> hash -> symbol\n"

  puts "Warming up JIT compiler (10,000 calls)..."
  10000.times do |i|
    dispatch_clj_impl(test_data[i % test_data.length], dispatch_table)
  end
  puts "Warmup complete.\n"

  puts "Running dispatch benchmark (3 iterations):"
  benchmark_dispatch(3, test_data, dispatch_table)

  puts "\n================================"
  puts "Benchmark Complete"
  puts "================================"
end

run_all_benchmarks(test_data, dispatch_table)

File: docs/caching/method_dispatch_bench.lua
#!/usr/bin/env lua
-- Micro-benchmark: generic function dispatch (Lua version)
-- Basic dispatch performance test using a dispatch table

-- Dispatch table
local dispatch_table = {
    ["number"] = function(x) return {"number", x} end,
    ["string"] = function(x) return {"string", #x} end,
    ["table"] = function(x) return {"table", #x} end,
    ["boolean"] = function(x) return {"boolean", x} end,
    ["other"] = function(x) return {"other", x} end,
}

local function get_type_key(x)
    local t = type(x)
    if t == "number" then
        return "number"
    elseif t == "string" then
        return "string"
    elseif t == "table" then
        return "table"
    elseif t == "boolean" then
        return "boolean"
    else
        return "other"
    end
end

local function dispatch_generic(x)
    local key = get_type_key(x)
    return dispatch_table[key](x)
end

-- Test data: 200,000 calls with 5-type repeating cycle
local test_data = {}
local cycle = {1, "test string", {1, 2, 3, 4, 5}, true, "symbol"}
for i = 1, 200000 do
    table.insert(test_data, cycle[((i - 1) % 5) + 1])
end

local function benchmark_dispatch(iterations)
    local times = {}
    for run = 1, iterations do
        local result = 0
        local start_time = os.clock()
        for _, item in ipairs(test_data) do
            local r = dispatch_generic(item)
            if r then result = result + 1 end
        end
        local elapsed = (os.clock() - start_time) * 1000.0
        print(string.format("  Run %d: %.1f seconds", run, elapsed / 1000.0))
        table.insert(times, result)
    end
    return times
end

local function run_all_benchmarks()
    print("\n================================")
    print("Lua Generic Function Dispatch Micro-Benchmark")
    print("================================")
    print(string.format("Implementation: Lua %s", _VERSION))
    print("Test data: 200,000 calls over repeating 5-type cycle")
    print("  Type cycle: number -> string -> table -> boolean -> other\n")

    print("Warming up interpreter (10,000 calls)...")
    for i = 1, 10000 do
        dispatch_generic(test_data[(i % 200000) + 1])
    end
    print("Warmup complete.\n")

    print("Running dispatch benchmark (3 iterations):")
    benchmark_dispatch(3)

    print("\n================================")
    print("Benchmark Complete")
    print("================================")
end

run_all_benchmarks()
